### PR TITLE
Background processing of ChannelManager and ChannelMonitor events

### PR DESCRIPTION
**Pull request by jkczyz**
Support processing `Event`s from `ChannelManager` and `ChainMonitor` in the background. Adds an `EventHandler` trait for users to implement and pass to `BackgroundProcessor`. Refactors the `EventsProvider` trait to take such an event handler.

This also addresses a race condition identified in #899 by making use of `PersistenceNotifierGuard` in `ChannelManager` when processing events.

### CODE REVIEW

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3548]: After #918 this probably should figure out if any events were processed and not always persist.

File: lightning/src/util/events.rs
Comment (TheBlueMatt) [inline lightning/src/util/events.rs:391]: This or the above needs a discussion of the locking semantics and reentrancy.

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [inline lightning/src/ln/channelmanager.rs:3555]: The code as written here will pass back the error of the first encountered error and stop processing events. However, any remaining events are lost because of the `drain`. So this leads to some other questions:
- Should event processing continue if handling one event fails?
- Can a user reasonably recover from an error while handling an event?
- Should unprocessed events remain pending for the next call?

The answer to these will determine what the interface will ultimately look like.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [inline lightning-background-processor/src/lib.rs:142]: There's an open question around how to handle errors in `BackgroundProcessor`. Currently, it will simply `unwrap` and fail if there is an error. Alternatively, it could return the error in the same manner as it does for the persister, which would necessitate defining an error type covering both.

**Review by jkczyz — COMMENTED** [review]
Added some comments that I could use some feedback on.

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3555]: Right, if we take an error from users we definitely shouldn't just drop events, we should keep feeding them to users. That said, I'm a little dubious that we should let users provide Errors either. We shouldn't take an error unless we can do something reasonable with it - maybe try the event again later or something, but short of that I'm not sure what an error even *means*.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:3555]: Yeah, I'm starting to think not allowing errors is the best approach. The user can always write an `EventHandler` to do whatever they want (including collecting errors or remaining events) and act upon it after `process_pending_events` returns. But, depending on how implemented, it could leave them in a state where `ChannelManager` is persisted without those events but the user hasn't "fully" handled them either. May just need to spell it out in the documentation.

That said, if they are using `BackgroundProcessor` there would be no way for them to examine the handler after processing.

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3555]: Right, I guess the only case where errors make sense is some kind of transitory issue that we know will resolve very soon, allowing us to handle the error momentarily while holding off persisting the ChannelManager. That said, I think you could just block in the handler in that case, as long as we only hold an event-specific lock?

Maybe we can have a separate lock allowing us to block any other event handlers from entering while still allowing adding new events from other things going on?

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:3555]: > Right, I guess the only case where errors make sense is some kind of transitory issue that we know will resolve very soon, allowing us to handle the error momentarily while holding off persisting the ChannelManager. That said, I think you could just block in the handler in that case, as long as we only hold an event-specific lock?
> 

Will update this to remove the return type from `handle_event`.

> Maybe we can have a separate lock allowing us to block any other event handlers from entering while still allowing adding new events from other things going on?

`pending_events` is already behind a `Mutex`. But I guess if we want to allow other event producers to add more events, we could immediately drop the lock after either draining fully into a `Vec` or swapping as before.

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs]: > pending_events is already behind a Mutex. But I guess if we want to allow other event producers to add more events, we could immediately drop the lock after either draining fully into a Vec or swapping as before.

Right, I guess I was thinking that if a user were handling events in parallel it would be a bug. Now that I think about it more maybe its fine, as long as that's what the user wants.

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: I think this meant to say #916.

File: lightning/src/ln/channelmanager.rs
Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: FWIW, swapping retains the same behavior as `get_and_clear_pending_events` since there the lock was held for the duration of the method and then the events were handled by the user. I updated the code accordingly.

File: lightning-background-processor/src/lib.rs
Comment (valentinewallace) [inline lightning-background-processor/src/lib.rs:321]: Wondering why we only call this on the last block, rather than for every block in `i..=ANTI_REORG_DELAY`?

**Comment by valentinewallace** [comment]
Could also be useful to document somewhere what we discussed offline re: expected behavior if we crash in the middle of handling a series of events.

**Review by jkczyz — COMMENTED** [review]
> Could also be useful to document somewhere what we discussed offline re: expected behavior if we crash in the middle of handling a series of events.

Done.

File: lightning/src/util/events.rs
Comment (TheBlueMatt) [inline lightning/src/util/events.rs:387]: Should we make it a part of the contract that callbacks into the provider are allowed/don't run locked, except for (possibly) serialization (in the case of a ChannelManager)? Its kinda hard to work with this API without that guarantee, but its kinda annoying to specify it.

File: lightning/src/chain/channelmonitor.rs
Comment (TheBlueMatt) [inline lightning/src/chain/channelmonitor.rs:229]: Maybe make the last four lines a comment instead of documentation? Feels weird to expose them.

File: lightning/src/util/events.rs
Comment (jkczyz) [reply lightning/src/util/events.rs:387]: For `ChannelManager`, both `process_pending_events` and `handle_event` take a read lock. The write lock is only taken after notification (i.e., after `process_pending_events` has returned) -- the handler itself doesn't serialize `ChannelManager`.

So I think the general advice is "don't introduce a deadlock". But now that I think of it, does `RwLock` allow reentrant reads? It appears so though the docs say it _might_ panic.

https://doc.rust-lang.org/std/sync/struct.RwLock.html#panics

File: lightning-net-tokio/src/lib.rs
Comment (jkczyz) [reply lightning-net-tokio/src/lib.rs:50]: Sure, I can remove it. But I don't quite understand why (if?) my change caused this to longer be necessary.

Comment (TheBlueMatt) [reply lightning-net-tokio/src/lib.rs:50]: The point of the channel here is to have something that you can use to wake up every time an Event may be generated (by hooking network reads). It's a stupid hack, but it works. Since we now do the handling based on the ChannelManager's own notification interface, I don't think there's any need to do it by hooking network reads anymore.

Comment (jkczyz) [reply lightning-net-tokio/src/lib.rs:50]: Ah, yeah, that's the piece I was missing. Sill need to call that interface here to avoid busy waiting.

File: lightning/src/util/events.rs
Comment (TheBlueMatt) [reply lightning/src/util/events.rs:387]: > So I think the general advice is "don't introduce a deadlock".

Right, but the docs don't *really* give you enough information to know what would and what would not cause a deadlock. Reading it naively, I'd think I need to queue all the events and then call back to ChannelManager separately as calling back to it may cause a deadlock. Providing more information would let users do more without fear, though we have to be precise and decide what we're willing to commit to in the API and what not.

> does RwLock allow reentrant reads?

I assumed it would...I mean you're allowed to take as many read locks as you want....

Comment (jkczyz) [reply lightning/src/util/events.rs:387]: > Right, but the docs don't _really_ give you enough information to know what would and what would not cause a deadlock. Reading it naively, I'd think I need to queue all the events and then call back to ChannelManager separately as calling back to it may cause a deadlock. Providing more information would let users do more without fear, though we have to be precise and decide what we're willing to commit to in the API and what not.

Understood, but these aren't `ChannelManager`'s docs. And note that this line is talking about someone implementing `EventsProvider` not `EventHandler`.  It's really up to an implementor of `EventsProvider` to determine what is safe for a handler to call. We don't need to restrict it at the trait level.

Comment (TheBlueMatt) [reply lightning/src/util/events.rs:387]: Right, I guess then the docs are somewhat confusing to me - I read them as "user, you should be careful of reentrancy when calling methods here", not "note to RL implementors, be careful of reentrancy". We should have some docs per-implementor that talks about reentrancy, I guess, though sadly Rust really hides the impl-level docs for a trait implementation, making them hard to find :(.

File: lightning-background-processor/src/lib.rs
Comment (valentinewallace) [reply lightning-background-processor/src/lib.rs:341]: Ah, I see. The comment above `open_channel!` may need updating since it implies that a ChannelManager will be persisted by the time it finishes. 

> Initially, I didn't move it, but the test failed in CI because open_channel! calls get_and_clear_pending_events, which races with BackgroundProcessor. That is it, the event handler may process the FundingGenerationReady event before open_channel! can do so, causing an assertion to fail in open_channel!.
 
Could we add a comment about this? :) Otherwise the test structure is a bit unintuitive to me

File: lightning/src/util/events.rs
Comment (valentinewallace) [inline lightning/src/util/events.rs:403]: Could point to the overall trait docs since those seem to have considerations (like the reentrancy) that could be pertinent as well

File: lightning-background-processor/src/lib.rs
Comment (valentinewallace) [inline lightning-background-processor/src/lib.rs:133]: Lol a casual 30-line method signature... Curious what the "best practice" for this would be. Definitely has been an issue since before this PR though.

File: lightning/src/chain/channelmonitor.rs
Comment (valentinewallace) [inline lightning/src/chain/channelmonitor.rs:229]: Could add or expose some public docs for this.

File: lightning/src/ln/channelmanager.rs
Comment (valentinewallace) [inline lightning/src/ln/channelmanager.rs:3585]: nit: space before todo + s/non-intuitive/unintuitive

**Review by valentinewallace — COMMENTED** [review]
Submitted some comments too early. Just some docs nits + the reentrancy clarification, otherwise LGTM

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3584]: While you're at it, can you add a persistence notifier in `get_and_clear_pending_msg_events` too, if it does actual work.

File: lightning/src/util/events.rs
Comment (jkczyz) [reply lightning/src/util/events.rs]: Re-wrote the docs to separate implementation and uses. Also added docs to `ChannelManager` and `ChainMonitor` implementations. Let me know if anything isn't clear or could be expanded upon.

Comment (jkczyz) [reply lightning/src/util/events.rs]: Rewrote as per earlier discussion. Let me know if anything could be cleared up further.

File: lightning-background-processor/src/lib.rs
Comment (jkczyz) [reply lightning-background-processor/src/lib.rs:133]: Not sure honestly unless we want to use trait objects. :(

File: lightning-net-tokio/src/lib.rs
Comment (jkczyz) [reply lightning-net-tokio/src/lib.rs:46]: Good question! I think `Cargo.lock` handles this for us since it will add the revision to the package source.

File: lightning/src/chain/channelmonitor.rs
Comment (valentinewallace) [reply lightning/src/chain/channelmonitor.rs:229]: Ah, I missed that, thanks

File: lightning/src/util/events.rs
Comment (TheBlueMatt) [reply lightning/src/util/events.rs]: Having a whole section dedicated to `Implementations` that contains instructions for how to implement just feels really wrong in public docs here? In practice, I can't imagine we ever expect users to implement this, so it becomes just a list of instructions to ourselves in public docs?

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3586]: These docs are collapsed by default on the ChannelManager page, beside four other trait implementations, making them basically invisible IMO. I don't know if there's a way to change rustdoc's behavior here, but I wonder if it makes sense to explicitly link to the known concrete implementations in the `EventsProvider` docs and say, eg, "eg see [\`ChannelManager::process_pending_events\`] and ..." which seems to handle opening the panel.

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3579]: Because message events aren't persisted as a part of ChannelManager (and we never assume that messages will be delivered between when we create them and give them to peers), I dont think this needs a persistence. However, we should be checking if a persist is required as a result of `process_pending_monitor_events` (both here and in `process_pending_events` below).

File: lightning-net-tokio/src/lib.rs
Comment (TheBlueMatt) [inline lightning-net-tokio/src/lib.rs:221]: Oops, I forgot that we're also relying on the event trigger thing here to call `peer_manager.process_events()` (which is also dumb...). Can you add a call to it after the select in the loop and maybe here after the disconnect as well? (while you're at it, we don't need the `peer_manager_ref` clone anymore, it seems?).

**Review by TheBlueMatt — COMMENTED** [review]
Aside from two functional comments here and the two doc comments, this all looks good.

File: lightning/src/util/events.rs
Comment (jkczyz) [reply lightning/src/util/events.rs]: Combined and rephrased in terms of requirements.

**Review by TheBlueMatt — APPROVED** [review]

File: lightning/src/chain/chainmonitor.rs
Comment (valentinewallace) [inline lightning/src/chain/chainmonitor.rs:320]: Should we note explicitly that calling back into the provider is OK?

**Review by valentinewallace — APPROVED** [review]

File: lightning/src/chain/chainmonitor.rs
Comment (jkczyz) [reply lightning/src/chain/chainmonitor.rs:320]: Hmm, good catch. Though we don't expect any need to call back, It's actually not safe to call `watch_channel` since it would attempt to take a `write` lock while already holding a `read` lock.

Comment (jkczyz) [reply lightning/src/chain/chainmonitor.rs:320]: Actually, never mind. We drop the read lock before calling the handler.

**Comment by TheBlueMatt** [comment]
D'oh. Sorry, I missed this conflicted with 851 before I hit merge. Could you do a quick rebase.

**Comment by jkczyz** [comment]
> D'oh. Sorry, I missed this conflicted with 851 before I hit merge. Could you do a quick rebase.

Sure thing. One sec.

**Review by valentinewallace — APPROVED** [review]

File: lightning/src/ln/channelmanager.rs
Comment (TheBlueMatt) [inline lightning/src/ln/channelmanager.rs:3706]: I think we need to check `check_free_holding_cells` for updates to persist as well.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:3706]: Ah, sorry. Didn't look at that closely enough. Figured it was a simple merge conflict. Let me fix it.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs:3706]: To confirm my understand, should we persist whenever `check_free_holding_cells` gets any failed HTLCs back from `maybe_free_holding_cell_htlcs`? Or is the condition more complicated?

https://github.com/rust-bitcoin/rust-lightning/blob/4e8bed20760ee1b53cf2fd15d4a9897780c4fab5/lightning/src/ln/channelmanager.rs#L3524

Comment (TheBlueMatt) [reply lightning/src/ln/channelmanager.rs:3706]: We should also persist if we get a `ChannelMonitorUpdate` back from `maybe_free_holding_cell_htlcs` (`ChannelMonitorUpdate`s are always a surefire reason to persist - if we're out of sync when we come back up we have to force-close). Reading the code in channel that should suffice, but maybe add a comment in `maybe_free_holding_cell_htlcs` and/or `free_holding_cell_htlcs` that notes that we won't persist unless we get some values back in the return.

Comment (jkczyz) [reply lightning/src/ln/channelmanager.rs]: Left the comment out since it's really a concern of `ChannelManager`. Made the logic in `check_free_holding_cells` explicit, though, so hopefully this is clear.

**Review by TheBlueMatt — APPROVED** [review]

**Pull request merged**
